apps/src/applab/dontMarshalApi.js

```diff
@@ -96,83 +154,219 @@ export function removeItem(array, index) {
   }
 }
 
-export function appendItem(array, item) {
-  dmapiValidateType('appendItem', 'list', array, 'array');
+function interpreterAppendItem(array, item) {
+  return appendItem(array, item, true);
+}
 
-  array.properties[array.length] = item;
-  array.length++;
-  return window.Applab.JSInterpreter.createPrimitive(array.length);
+export function appendItem(array, item, calledWithinInterpreter) {
+  if (calledWithinInterpreter) {
+    dmapiValidateType('appendItem', 'list', array, 'array');
+  } else {
+    apiValidateType({}, 'appendItem', 'list', array, 'array');
+  }
+
+  const arrayValues = calledWithinInterpreter ? array.properties : array;
+
+  arrayValues[array.length] = item;
+  if (calledWithinInterpreter) {
+    // In the interpreter, we must manually update the length property:
+    array.length++;
+    // And we must create an interpreter primitive to wrap the return value:
+    return window.Applab.JSInterpreter.createPrimitive(array.length);
+  } else {
+    return array.length;
+  }
 }
 
 // ImageData RGB helper functions
 
 // TODO: more parameter validation (data array type, length), error output
 
-export function getRed(imageData, x, y) {
-  if (imageData.properties.data && imageData.properties.width) {
-    var pixelOffset = y * imageData.properties.width * 4 + x * 4;
-    return imageData.properties.data.properties[pixelOffset].toNumber();
+function interpreterGetRed(imageData, x, y) {
+  return getRed(imageData, x, y, true);
+}
+
+export function getRed(imageData, x, y, calledWithinInterpreter) {
+  const imageDataProperties = calledWithinInterpreter
+    ? imageData.properties
+    : imageData;
+  if (imageDataProperties.data && imageDataProperties.width) {
+    const pixelOffset = y * imageDataProperties.width * 4 + x * 4;
+    const redOffset = pixelOffset;
+    return calledWithinInterpreter
+      ? imageDataProperties.data.properties[redOffset].toNumber()
+      : imageDataProperties.data[redOffset];
```

Comment:
There's a lot of duplicated code here. I'd like to try to find a way to remove some of the duplication. @joshlory and I were talking about this a bit offline, and we came up with a couple ideas.
1. Create a wrapper that passes input.properties if called within the interpreter. i.e.
```
wrap = (fn) => {
  return (...args) => {
    args.forEach((arg, n) =>
      if (arg.properties) {
        args[n] = arg.properties
      }
    );
    return fn.apply(this, args);
  }
}
```
2. Handle this further upstream so we pass the correct properties to functions like getRed and dontMarshalApi doesn't need to know the difference.
3. Refactor within JS-Interpreter to use a version that works for both native and JS-Interpreter (this is more for the array manipulation functions)

Do you know why we need to use arg.properties in so many places here?

Reply:
@jmkulwik @joshlory Can you be clear on which duplication you'd like to avoid? It is possible to make all of the `getRed()` and `setRed()` type functions wrap a simpler template, but I'm not sure that's what you're going for.

In response to your ideas:

(1) - doesn't handle any of the complex scenarios (like nested objects or arrays) or the need to create primitives in the response (things that are normally taken care of by the marshaling code)
(2) - this is essentially what marshaling does. we do that for something like 98% of our APIs. These are the exceptions because they either modify parameters (in/out params) [list functions], need high perf without marshaling huge arrays [get imagedata APIs], or both [set imagedata APIs]
(3) - for the list functions, I have some question as to why they were built this way. I was interested in making them simple wrappers of existing array functions (e.g. these could all be calls to `Array.splice()` which already has 2 implementations - one for interpreter, one built-in for normal JS). The only reason I didn't do this was because we've added strange behaviors (e.g. rounding the index param in case it is floating point?)

I can look at (3), but for now, but the fixes above are legit and the code as-is is completely broken. I'd be happy to get fixes merged in and then discuss improved fixes that are cleaner. Believe me, I thought about it for a little while. It's not a super simple problem.

Reply:
Comments left by @joshlory in Slack while Github was down:

> It's annoying that we created "don't marshal" APIs in the first place.

Is it written down anywhere _why_ we created this in the first place? (I think perf?) Looking back now, how do you think we should've done it instead?

Would it be possible to move `insertItem`, `appendItem` and `removeItem` out of dontMarshalApi.js and into the normal API?

Reply:
To expand: what's the difference between "custom marshal" and "don't marshal". Can we update marshaling in general to treat these data types as special, so we don't need to special-case the APIs that pass them as arguments or return values?

Reply:
We can't move them out of `dontMarshalApi.js` because they modify the arrays that are passed as a parameter to them. Our default marshaling is "by value". Arguments/parameters are "copied" on the way in, and the return value is "copied" on the way out.

"custom marshal" allows us to essentially keep objects in the interpreter as thin wrappers around data that is actually maintained outside the interpreter. As an example, a `Sprite` in gamelab is always custom marshaled. When it is passed as a parameter, we don't copy it (by value) from "native" to "interpreted", we create a interpreter wrapper (essentially passing it by reference).

"don't marshal" is used in applab for higher-performance scenarios (ImageData) or where we want to modify a parameter (since marshaling copies, changes made in the copy within the native function have no impact in the interpreted world)

Reply:
Should we eventually converge the "don't marshal" and "custom marshal" code paths?